ccron/src/periodic_action.cpp

```diff
@@ -0,0 +1,79 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the sub-component's license, as noted in the LICENSE
+// file.
+
+#include "ccron/periodic_action.hpp"
+
+#include "ccron_msgs.cmf.hpp"
+
+#include <cstdint>
+#include <mutex>
+#include <utility>
+#include <vector>
+
+namespace concord::cron {
+
+static PeriodicActionSchedule schedule;
+static std::once_flag init_once;
+
+static const auto periodicCronReservedPageId = 3;
+
+static void saveSchedule(bftEngine::IReservedPages& reserved_pages) {
+  auto output = std::vector<std::uint8_t>{};
+  serialize(output, schedule);
+  // TODO: reserved page ID
+  reserved_pages.saveReservedPage(
+      periodicCronReservedPageId, output.size(), reinterpret_cast<const char*>(output.data()));
+}
+
+CronEntry periodicAction(std::uint32_t position,
+                         const Action& action,
+                         const std::chrono::milliseconds& period,
+                         bftEngine::IReservedPages& reserved_pages) {
+  std::call_once(init_once, [&reserved_pages]() {
+    auto input = std::vector<std::uint8_t>(reserved_pages.sizeOfReservedPage());
+    if (reserved_pages.loadReservedPage(
+            periodicCronReservedPageId, input.size(), reinterpret_cast<char*>(input.data()))) {
+      deserialize(input, schedule);
+    }
+  });
+
+  auto rule = [position](const Tick& tick) {
+    auto component_it = schedule.components.find(tick.component_id);
+    if (component_it == schedule.components.cend()) {
+      return true;
+    }
+    auto& component_table = component_it->second;
+    auto next_invocation_it = component_table.find(position);
+    // If the position doesn't exist in the persisted schedule (i.e. the table was created with new positions after
+    // persistance), assume this is the first invocation and allow it.
+    // If the position exists, check the time schedule.
+    if (next_invocation_it == component_table.cend() ||
+        tick.ms_since_epoch.count() >= next_invocation_it->second.ms_since_epoch) {
+      return true;
+    }
+    return false;
+  };
+
+  auto schedule_next = [position, &reserved_pages, period](const Tick& tick) {
+    auto next = NextInvocation{(tick.ms_since_epoch + period).count()};
+    schedule.components[tick.component_id][position] = next;
+    saveSchedule(reserved_pages);
```

Comment:
Yes, good point. Maybe one thing we can do is to insert a special entry with an always `true` rule, empty action and an `on_schedule` that saves the page at the end of the table on every `periodicAction()` call. Also, user entries will be without an `on_schedule`. The question is how to make the last position in the table special. Let me think about it a bit more and will implement a solution.

Reply:
I have implemented as a special persistence cron table entry that is always the last once, after all periodic action entries. We then accumulate the changes in memory and persist them at the end once, only if something has been scheduled.

Copying the comment from the code to make it easier to read:

```
// Create a cron table entry that represents a periodic `action` that occurs every `period` milliseconds.
// Works in conjunction with the `persistPeriodicSchedule()` entry:
//  -----------------------------------------------
//  | position|rule|action|schedule_next|on_remove|
//  -----------------------------------------------
//  |    1    | R1 |  A1  |     S1      |   RM1   |
//  |    2    | R2 |  A2  |     S2      |   RM2   |
//  |.............................................|
//  |    P    |  ~ |   ~  |     ~       |    ~    |
//  -----------------------------------------------
// Position P is the position of the `persistPeriodicSchedule()` entry that is run `after` all other periodic action
// entries. Rationale is that we want to avoid serializing the schedule to reserved pages multiple times (for every
// periodic action in the table) per tick. Instead, we accumulate the changes in memory and then use the
// `persistPeriodicSchedule()` entry to persist once, at the end.
CronEntry periodicAction(std::uint32_t position,
                         const Action& action,
                         const std::chrono::milliseconds& period,
                         bftEngine::IReservedPages& reserved_pages);

// Create a cron entry that persists the periodic actions schedule. To work as expected, users must ensure that its
// position is bigger than the positions of all other periodic action entries (see above).
CronEntry persistPeriodicSchedule(std::uint32_t position, bftEngine::IReservedPages& reserved_pages);
```

Reply:
Very cool. Thanks!